File: job-modeled-toolkits/29-expense-audit-toolkit/02-expense-review/src/audit.js
/* Expense audit logic for the review app.
 *
 * Pure functions: validate an expense, flag it against the policy, and total a
 * run, with no DOM access. The interface in ui.js wires these to the review
 * queue; the test harness in tests.js checks them. The logic mirrors the Python
 * auditor in 01 to the cent, with money held in integer cents.
 *
 * Runs in the browser (attaches ExpenseAudit to the window) and under Node (it
 * exports the same object), so the tests run in either place.
 */
(function (global) {
  "use strict";

  var MILEAGE = "Mileage";
  var YES_NO = { yes: true, no: false };

  function toCents(value) {
    var text = String(value).trim();
    var negative = text.charAt(0) === "-";
    if (negative) { text = text.slice(1); }
    var parts = text.split(".");
    var whole = parts[0] === "" ? 0 : parseInt(parts[0], 10);
    var fraction = (parts[1] || "").slice(0, 3);
    while (fraction.length < 3) { fraction += "0"; }
    var thousandths = parseInt(fraction, 10);
    var cents = whole * 100 + Math.floor(thousandths / 10);
    if (thousandths % 10 >= 5) { cents += 1; }
    return negative ? -cents : cents;
  }

  function formatMoney(cents) {
    return (cents / 100).toLocaleString("en-CA", { style: "currency", currency: "CAD" });
  }

  function buildPolicy(raw) {
    var caps = {};
    Object.keys(raw.caps || {}).forEach(function (cat) { caps[cat] = toCents(raw.caps[cat]); });
    return {
      mileageRateCents: toCents(raw.mileage_rate),
      receiptThresholdCents: toCents(raw.receipt_threshold),
      caps: caps,
    };
  }

  function allowedCategories(policy) {
    return Object.keys(policy.caps).concat([MILEAGE]);
  }

  function validateExpense(raw, policy) {
    var id = String(raw.expense_id || "").trim() || "(missing id)";
    var label = "Expense " + id;
    var required = ["expense_id", "date", "employee", "category", "amount", "receipt"];
    for (var i = 0; i < required.length; i++) {
      if (String(raw[required[i]] === undefined ? "" : raw[required[i]]).trim() === "") {
        return { ok: false, error: label + ": missing " + required[i] };
      }
    }
    var category = String(raw.category).trim();
    if (allowedCategories(policy).indexOf(category) === -1) {
      return { ok: false, error: label + ": category " + category + " has no policy" };
    }
    if (!/^\d{4}-\d{2}-\d{2}$/.test(String(raw.date).trim())) {
      return { ok: false, error: label + ": date must be YYYY-MM-DD" };
    }
    var amountCents = toCents(raw.amount);
    if (amountCents <= 0) { return { ok: false, error: label + ": amount must be above zero" }; }
    var kmText = String(raw.km === undefined ? "" : raw.km).trim();
    var km = kmText === "" ? 0 : Number(kmText);
    if (!(km >= 0)) { return { ok: false, error: label + ": km cannot be negative" }; }
    if (category === MILEAGE && km <= 0) {
      return { ok: false, error: label + ": a mileage claim needs kilometres above zero" };
    }
    var receiptText = String(raw.receipt).trim().toLowerCase();
    if (!(receiptText in YES_NO)) { return { ok: false, error: label + ": receipt must be yes or no" }; }
    return {
      ok: true,
      value: {
        expense_id: id, date: String(raw.date).trim(), employee: String(raw.employee).trim(),
        category: category, amountCents: amountCents, km: km, receipt: YES_NO[receiptText],
      },
    };
  }

  function mileageAmountCents(km, rateCents) {
    return Math.round(km * rateCents);
  }

  function computedAmountCents(expense, policy) {
    return expense.category === MILEAGE
      ? mileageAmountCents(expense.km, policy.mileageRateCents)
      : expense.amountCents;
  }

  function duplicateKeys(expenses) {
    var counts = {};
    expenses.forEach(function (e) {
      var key = [e.employee, e.date, e.category, e.amountCents].join("|");
      counts[key] = (counts[key] || 0) + 1;
    });
    var dup = {};
    Object.keys(counts).forEach(function (k) { if (counts[k] > 1) { dup[k] = true; } });
    return dup;
  }

  function flagsFor(expense, policy, dupKeys) {
    var flags = [];
    if (expense.category === MILEAGE) {
      if (expense.amountCents !== mileageAmountCents(expense.km, policy.mileageRateCents)) {
        flags.push("MILEAGE_MISMATCH");
      }
    } else {
      var cap = policy.caps[expense.category];
      if (cap !== undefined && expense.amountCents > cap) { flags.push("OVER_CAP"); }
      if (expense.amountCents > policy.receiptThresholdCents && !expense.receipt) { flags.push("NO_RECEIPT"); }
    }
    var key = [expense.employee, expense.date, expense.category, expense.amountCents].join("|");
    if (dupKeys[key]) { flags.push("DUPLICATE"); }
    return flags;
  }

  function auditAll(expenses, policy) {
    var dupKeys = duplicateKeys(expenses);
    var rows = expenses.map(function (e) {
      var flags = flagsFor(e, policy, dupKeys);
      return {
        expense_id: e.expense_id, date: e.date, employee: e.employee, category: e.category,
        amountCents: e.amountCents, km: e.km, receipt: e.receipt,
        computedCents: computedAmountCents(e, policy), flags: flags,
        status: flags.length ? "Flagged" : "Approved",
      };
    });
    var flagged = rows.filter(function (r) { return r.flags.length; });
    function sum(list) { return list.reduce(function (a, r) { return a + r.amountCents; }, 0); }
    var totals = {
      totalClaimed: sum(rows),
      flaggedAmount: sum(flagged),
      approvedAmount: sum(rows.filter(function (r) { return !r.flags.length; })),
      flaggedCount: flagged.length,
      approvedCount: rows.length - flagged.length,
      overCapCount: rows.filter(function (r) { return r.flags.indexOf("OVER_CAP") !== -1; }).length,
      noReceiptCount: rows.filter(function (r) { return r.flags.indexOf("NO_RECEIPT") !== -1; }).length,
      duplicateCount: rows.filter(function (r) { return r.flags.indexOf("DUPLICATE") !== -1; }).length,
      mileageMismatchCount: rows.filter(function (r) { return r.flags.indexOf("MILEAGE_MISMATCH") !== -1; }).length,
    };
    return { rows: rows, totals: totals };
  }

  var api = {
    toCents: toCents, formatMoney: formatMoney, buildPolicy: buildPolicy,
    validateExpense: validateExpense, mileageAmountCents: mileageAmountCents,
    computedAmountCents: computedAmountCents, duplicateKeys: duplicateKeys,
    flagsFor: flagsFor, auditAll: auditAll,
  };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = api;
  } else {
    global.ExpenseAudit = api;
  }
})(typeof window !== "undefined" ? window : globalThis);

File: job-modeled-toolkits/29-expense-audit-toolkit/01-expense-auditor/audit.py
"""Expense and travel policy audit logic.

Pure functions only. Each takes plain values and returns plain values, with no
file or console access. The CLI layer in cli.py reads and writes; the browser
review app in 02 mirrors this same logic to the cent.

Each expense line is checked against the policy and given zero or more flags:

  MILEAGE_MISMATCH - a mileage claim whose amount does not equal kilometres times
                     the prescribed rate.
  OVER_CAP         - a category amount above its daily cap.
  NO_RECEIPT       - an amount above the receipt threshold with no receipt attached.
  DUPLICATE        - the same employee, date, category, and amount appearing more
                     than once.

A line with no flags is approved; a flagged line goes to the review queue.

All money is decimal.Decimal rounded half up to the cent.
"""

from decimal import Decimal, ROUND_HALF_UP

CENT = Decimal("0.01")
MILEAGE_CATEGORY = "Mileage"


def money(value):
    return value.quantize(CENT, rounding=ROUND_HALF_UP)


def mileage_amount(km, rate):
    """The allowed amount for a mileage claim: kilometres times the rate."""
    return money(km * rate)


def duplicate_keys(expenses):
    """The set of (employee, date, category, amount) keys that appear more than once."""
    counts = {}
    for e in expenses:
        key = (e["employee"], e["date"], e["category"], e["amount"])
        counts[key] = counts.get(key, 0) + 1
    return {key for key, count in counts.items() if count > 1}


def computed_amount(expense, policy):
    """The amount the policy expects: the mileage formula for mileage, the claimed
    amount otherwise."""
    if expense["category"] == MILEAGE_CATEGORY:
        return mileage_amount(expense["km"], policy["mileage_rate"])
    return money(expense["amount"])


def flags_for(expense, policy, dup_keys):
    """The list of policy flags on one expense, in a stable order."""
    flags = []
    category = expense["category"]
    amount = money(expense["amount"])

    if category == MILEAGE_CATEGORY:
        if amount != mileage_amount(expense["km"], policy["mileage_rate"]):
            flags.append("MILEAGE_MISMATCH")
    else:
        cap = policy["caps"].get(category)
        if cap is not None and amount > cap:
            flags.append("OVER_CAP")
        if amount > policy["receipt_threshold"] and not expense["receipt"]:
            flags.append("NO_RECEIPT")

    key = (expense["employee"], expense["date"], category, expense["amount"])
    if key in dup_keys:
        flags.append("DUPLICATE")
    return flags


def audit(expenses, policy):
    """Audit every expense and return the rows plus the run totals."""
    dup_keys = duplicate_keys(expenses)
    rows = []
    for e in expenses:
        flags = flags_for(e, policy, dup_keys)
        rows.append({
            "expense_id": e["expense_id"],
            "date": e["date"],
            "employee": e["employee"],
            "category": e["category"],
            "amount": money(e["amount"]),
            "km": e["km"],
            "receipt": "yes" if e["receipt"] else "no",
            "computed_amount": computed_amount(e, policy),
            "flags": flags,
            "status": "Approved" if not flags else "Flagged",
        })

    flagged = [r for r in rows if r["flags"]]
    totals = {
        "total_claimed": money(sum((r["amount"] for r in rows), Decimal("0.00"))),
        "flagged_amount": money(sum((r["amount"] for r in flagged), Decimal("0.00"))),
        "approved_amount": money(sum((r["amount"] for r in rows if not r["flags"]), Decimal("0.00"))),
        "flagged_count": len(flagged),
        "approved_count": len(rows) - len(flagged),
        "over_cap_count": sum(1 for r in rows if "OVER_CAP" in r["flags"]),
        "no_receipt_count": sum(1 for r in rows if "NO_RECEIPT" in r["flags"]),
        "duplicate_count": sum(1 for r in rows if "DUPLICATE" in r["flags"]),
        "mileage_mismatch_count": sum(1 for r in rows if "MILEAGE_MISMATCH" in r["flags"]),
    }
    return {"rows": rows, "totals": totals}
